common/src/primitives/merkle/tree.rs

```diff
@@ -0,0 +1,560 @@
+    if n == 0 {
+        return 1;
+    }
+    let leading_zeros = (n - 1).leading_zeros() as usize;
+    let active_bits = usize::BITS as usize - leading_zeros;
+    (1 << active_bits) as usize
+}
+
+fn create_merkletree_padding(elements: &[H256]) -> Vec<H256> {
+    let orig_size = elements.len();
+    let pow2_size = next_pow2(orig_size);
+
+    assert!(pow2_size >= orig_size);
+
+    let mut padding = Vec::with_capacity(pow2_size - orig_size);
+    for _idx in orig_size..pow2_size {
+        let to_hash = padding
+            .last()
+            .unwrap_or_else(|| elements.last().expect("We already checked it's not empty"));
+        let to_push = default_hash(to_hash);
+        padding.push(to_push);
+    }
+    padding
+}
+
+impl MerkleTree {
+    fn combine_pair(left: &H256, right: &H256) -> H256 {
+        let mut hasher = DefaultHashAlgoStream::new();
+        hasher.write(left.as_bytes());
+        hasher.write(right.as_bytes());
+        H256::from(hasher.finalize())
+    }
+
+    /// Create a new merkle tree from a list of leaves.
+    pub fn from_leaves(leaves: Vec<H256>) -> Result<Self, MerkleTreeFormError> {
+        // TODO: separate padding from this function and create a type that includes padding
+        //       on creation by taking Vec<H256> as input and padding it and wrapping it with a strong type
+        if leaves.is_empty() {
+            return Err(MerkleTreeFormError::TooSmall(leaves.len()));
+        }
+        let padding = create_merkletree_padding(&leaves);
+        let leaves = leaves.into_iter().chain(padding).collect::<Vec<_>>();
+        let steps = leaves.len() - 1;
+        let mut tree = Vec::with_capacity(2 * leaves.len() - 1);
+        tree.extend(leaves.into_iter());
+        for i in 0..steps {
+            let el = Self::combine_pair(&tree[i * 2], &tree[i * 2 + 1]);
+            tree.push(el);
+        }
+        let res = Self { tree };
+        Ok(res)
+    }
+
+    /// Get the root of the merkle tree.
+    pub fn root(&self) -> H256 {
+        *self.tree.last().expect("By design, at least one element must exist")
+    }
+
+    fn leaves_count_from_tree_size(
+        tree_size: NonZeroUsize,
+    ) -> Result<NonZeroUsize, MerkleTreeProofExtractionError> {
+        if (tree_size.get() + 1).count_ones() != 1 {
+            return Err(MerkleTreeProofExtractionError::InvalidTreeSize(
+                tree_size.get(),
+            ));
+        }
+
+        let tree_size = tree_size.get();
+        let leaves_count = (tree_size + 1) >> 1;
+        debug_assert!(leaves_count.is_power_of_two());
+        Ok(NonZeroUsize::new(leaves_count).expect("By design, tree_size is always > 0"))
+    }
+
+    /// Ensure the leaves indices are sorted and unique
+    fn is_sorted_and_unique(leaves_indices: &[u32]) -> bool {
+        let vals: BTreeSet<u32> = leaves_indices.iter().copied().collect();
+        let vals: Vec<u32> = vals.into_iter().collect();
+        vals == leaves_indices
+    }
```

Comment:
Seems a bit overkill to construct these intermediate data structures. I think we can do this with iterators alone. Very roughly something like this should do:

```rust
indices.tuple_windows::<(_, _)>().all(|(i, j)| i < j)
```

Reply:
Nice idea. Done.